synapse/storage/databases/main/cache.py

```diff
@@ -554,3 +577,103 @@ def get_cache_stream_token_for_writer(self, instance_name: str) -> int:
             return self._cache_id_gen.get_current_token_for_writer(instance_name)
         else:
             return 0
+
+    def _clean_up_cache_invalidation_wrapper(self) -> None:
+        async def _clean_up_cache_invalidation_background():
+            delete_up_to: int = (
+                self.hs.get_clock().time_msec()
+                - RETENTION_PERIOD_OF_CACHE_INVALIDATIONS_MILLISEC
+            )
+
+            in_backlog = await self._clean_up_batch_of_old_cache_invalidations(
+                delete_up_to
+            )
+
+            # Vary how long we wait before calling again depending on whether we
+            # are still sifting through backlog or we have caught up.
+            if in_backlog:
+                next_interval = CATCH_UP_CLEANUP_INTERVAL_MILLISEC
+            else:
+                next_interval = REGULAR_CLEANUP_INTERVAL_MILLISEC
+
+            self.hs.get_clock().call_later(
+                next_interval, self._clean_up_cache_invalidation_wrapper
+            )
+
+        run_as_background_process(
+            "clean_up_old_cache_invalidations", _clean_up_cache_invalidation_background
+        )
+
+    async def _clean_up_batch_of_old_cache_invalidations(
+        self, delete_up_to_millisec: int
+    ) -> bool:
+        """
+        Cleans up a batch of old cache invalidations.
+
+        Up to `CLEAN_UP_BATCH_SIZE` rows will be deleted at once.
+
+        Returns true iff we were limited by batch size (i.e. we are in backlog).
+        """
+
+        def _clean_up_batch_of_old_cache_invalidations_txn(
+            txn: LoggingTransaction,
+        ) -> bool:
+            # First get the earliest stream ID
+            txn.execute(
+                """
+                SELECT stream_id FROM cache_invalidation_stream_by_instance
+                ORDER BY stream_id LIMIT 1
```

Comment:
Is it pretty standard to omit the order if it's ascending?

I had to think about at first glance to make sure this matches the comment

```suggestion
                ORDER BY stream_id ASC
                LIMIT 1
```

Reply:
I'm ambivalent about it. It is kind of 'obvious' to me at least that ordering is ascending by default, so I don't even think about adding a keyword unless it's `DESC` most of the time. But I don't mind either way